x/bridge/keeper/keeper_test.go

```diff
@@ -0,0 +1,156 @@
+	newParams := s.GetParams()
+	newParams.Signers = append(newParams.Signers, signers...)
+
+	_, err := s.msgServer.UpdateParams(s.Ctx, &types.MsgUpdateParams{
+		Sender:    s.authority,
+		NewParams: newParams,
+	})
+	s.Require().NoError(err)
+}
+
+func (s *KeeperTestSuite) EnableAssets(assetIDs ...types.AssetID) {
+	for _, assetID := range assetIDs {
+		_, err := s.msgServer.ChangeAssetStatus(s.Ctx, &types.MsgChangeAssetStatus{
+			Sender:    s.authority,
+			AssetId:   assetID,
+			NewStatus: types.AssetStatus_ASSET_STATUS_OK,
+		})
+		s.Require().NoError(err)
+	}
+}
+
+func (s *KeeperTestSuite) SetVotesNeeded(votesNeeded uint64) {
+	newParams := s.GetParams()
+	newParams.VotesNeeded = votesNeeded
+
+	_, err := s.msgServer.UpdateParams(s.Ctx, &types.MsgUpdateParams{
+		Sender:    s.authority,
+		NewParams: newParams,
+	})
+	s.Require().NoError(err)
+}
+
+func (s *KeeperTestSuite) GetParams() types.Params {
+	resp, err := s.queryClient.Params(s.Ctx, new(types.QueryParamsRequest))
+	s.Require().NoError(err)
+	return resp.GetParams()
+}
+
+func (s *KeeperTestSuite) GetLastTransferHeight(assetID types.AssetID) uint64 {
+	resp, err := s.queryClient.LastTransferHeight(s.Ctx, &types.LastTransferHeightRequest{
+		AssetId: assetID,
+	})
+	s.Require().NoError(err)
+	return resp.LastTransferHeight
+}
+
+func (s *KeeperTestSuite) GetModuleAddress() string {
+	return s.App.AccountKeeper.GetModuleAddress(types.ModuleName).String()
+}
+
+// GetBridgeTFDenoms returns list of denoms created by the bridge module.
+// Demons are requested from the tokenfactory.
+func (s *KeeperTestSuite) GetBridgeTFDenoms() []string {
+	req := &tokenfactorytypes.QueryDenomsFromCreatorRequest{
+		Creator: s.GetModuleAddress(),
+	}
+	resp, err := s.tfQueryClient.DenomsFromCreator(s.Ctx, req)
+	s.Require().NoError(err)
+	return resp.Denoms
+}
+
+// GetBridgeDenoms generates list of denoms based on the assets module param.
+func (s *KeeperTestSuite) GetBridgeDenoms() []string {
+	params := s.GetParams()
+	bridgeDenoms := make([]string, 0, len(params.Assets))
+	moduleAddr := s.GetModuleAddress()
+	for _, asset := range params.Assets {
+		denom, err := tokenfactorytypes.GetTokenDenom(moduleAddr, asset.Name())
+		s.Require().NoError(err)
+		bridgeDenoms = append(bridgeDenoms, denom)
+	}
+	return bridgeDenoms
+}
+
+func (s *KeeperTestSuite) GetTFDenom(assetID types.AssetID) string {
+	moduleAddr := s.GetModuleAddress()
+	denom, err := tokenfactorytypes.GetTokenDenom(moduleAddr, assetID.Name())
+	s.Require().NoError(err)
+	return denom
+}
+
+func (s *KeeperTestSuite) GetAddrFromBech32(addr string) sdk.AccAddress {
+	result, err := sdk.AccAddressFromBech32(addr)
+	s.Require().NoError(err)
+	return result
+}
+
+func (s *KeeperTestSuite) HasEvent(eventType string) bool {
+	events := s.Ctx.EventManager().Events()
+	const eventIdxNotFound = -1
+	eventIdx := slices.IndexFunc(events, func(e sdk.Event) bool {
+		return e.Type == eventType
+	})
+	return eventIdx != eventIdxNotFound
+}
```

Comment:
Its good we are checking events in the tests, just want to highlight that we cannot use events as something we trigger actions off of